latch_cli/docker_utils/__init__.py

```diff
@@ -0,0 +1,210 @@
+                ],
+                order=DockerCmdBlockOrder.precopy,
+            ),
+            DockerCmdBlock(
+                comment="build and configure conda environment",
+                commands=[
+                    f"copy {conda_env_file_name} /opt/latch/{conda_env_file_name}",
+                    f"run conda env create --file /opt/latch/{conda_env_file_name} --name {env_name}",
+                    f"""shell ["conda", "run", "--name", "{env_name}", "/bin/bash", "-c"]""",
+                    "run /opt/conda/bin/pip install --upgrade latch",
+                ],
+                order=DockerCmdBlockOrder.precopy,
+            ),
+        ]
+
+    if (pkg_root / "setup.py").exists() or (pkg_root / "pyproject.toml").exists():
+        print("Adding local package installation phase")
+        commands.append(
+            DockerCmdBlock(
+                comment="add local package to python environment",
+                commands=["run pip install --editable /root/"],
+                order=DockerCmdBlockOrder.postcopy,
+            )
+        )
+
+    if (pkg_root / "requirements.txt").exists():
+        print("Adding python package installation phase")
+        commands.append(
+            DockerCmdBlock(
+                comment="add requirements from `requirements.txt` to python environment",
+                commands=[
+                    "copy requirements.txt /opt/latch/requirements.txt",
+                    "pip install --requirement /opt/latch/requirements.txt",
+                ],
+                order=DockerCmdBlockOrder.precopy,
+            )
+        )
+
+    return commands
+
+
+def generate_dockerfile(pkg_root: Path, outfile: Path) -> None:
+    """Generate a best effort Dockerfile from files in the workflow directory.
+
+    Args:
+        pkg_root: A path to a workflow directory.
+        outfile: The path to write the generated Dockerfile.
+
+    Example:
+
+        >>> generate_dockerfile(Path("test-workflow"), Path("test-workflow/Dockerfile"))
+            # The resulting file structure will look like
+            #   test-workflow
+            #   ├── Dockerfile
+            #   └── ...
+    """
+
+    print("Generating Dockerfile")
+    try:
+        with open(pkg_root / latch_constants.pkg_config) as f:
+            config: LatchWorkflowConfig = LatchWorkflowConfig(**json.load(f))
+            print("  - base image:", config.base_image)
+            print("  - latch version:", config.latch_version)
+    except FileNotFoundError as e:
+        raise RuntimeError(
+            "Could not find .latch file in supplied directory. If your workflow was created previously to release 2.13.0, you may need to run `latch init` to generate a .latch file."
+        ) from e
+
+    with outfile.open("w") as f:
+        f.write("\n".join(get_prologue(config)) + "\n\n")
+
+        commands = infer_commands(pkg_root)
+        pre_commands = [c for c in commands if c.order == DockerCmdBlockOrder.precopy]
```

Comment:
doesn't really make sense to pre-filter
you could just put an `if` in a loop over all commands lol